File: brainplus-engine/src/api/openai_compat.rs
//! OpenAI 兼容 API 流式客户端 (DeepSeek/Qwen/OpenRouter/Groq/Mistral/...)
//! 对齐 src/lib/api/openai-compat.ts

use super::{ChatMessage, ProviderConfig, StreamEvent, ToolDef, Usage};
use serde_json::Value;

pub async fn stream_openai_compat(
    config: &ProviderConfig,
    messages: &[ChatMessage],
    tools: &[ToolDef],
    system_prompt: Option<&str>,
    max_tokens: Option<u32>,
    tx: tokio::sync::mpsc::Sender<StreamEvent>,
) -> Result<Usage, String> {
    let base_url = config.base_url.as_deref().unwrap_or("https://api.deepseek.com");
    let url = format!("{}/v1/chat/completions", base_url.trim_end_matches('/'));

    let mut openai_msgs: Vec<Value> = Vec::new();

    if let Some(sys) = system_prompt {
        openai_msgs.push(serde_json::json!({"role": "system", "content": sys}));
    }

    for m in messages {
        if m.role == "system" { continue; }
        let mut obj = serde_json::json!({"role": m.role, "content": m.content});
        if let Some(ref rc) = m.reasoning_content { obj["reasoning_content"] = serde_json::json!(rc); }

        if let Some(ref tc) = m.tool_calls {
            obj["tool_calls"] = serde_json::json!(tc.iter().map(|c| {
                serde_json::json!({
                    "id": c.id,
                    "type": "function",
                    "function": { "name": c.name, "arguments": serde_json::to_string(&c.input).unwrap_or_default() },
                })
            }).collect::<Vec<_>>());
        }
        // role=tool 必须有 tool_call_id（OpenAI/DeepSeek 强制要求）
        if m.role == "tool" || m.tool_call_id.is_some() {
            obj["role"] = serde_json::Value::String("tool".into());
            obj["tool_call_id"] = serde_json::json!(m.tool_call_id.as_deref().unwrap_or("unknown"));
        }

        openai_msgs.push(obj);
    }

    let openai_tools: Vec<Value> = tools.iter().map(|t| {
        serde_json::json!({
            "type": "function",
            "function": {
                "name": t.name,
                "description": t.description,
                "parameters": t.input_schema,
            },
        })
    }).collect();

    let mut body = serde_json::json!({
        "model": config.model_id,
        "messages": openai_msgs,
        "max_tokens": max_tokens.unwrap_or(4096),
        "stream": true,
    });
    if !openai_tools.is_empty() {
        body["tools"] = serde_json::json!(openai_tools);
    }

    let client = reqwest::Client::new();
    let response = client.post(&url)
        .header("Authorization", format!("Bearer {}", config.api_key))
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("API 请求失败: {e}"))?;

    let status = response.status();
    if !status.is_success() {
        let err_body = response.text().await.unwrap_or_default();
        return Err(format!("API HTTP {}: {}", status.as_u16(), &err_body[..300.min(err_body.len())]));
    }

    let mut stream = response.bytes_stream();
    use futures::StreamExt;
    let mut buffer = String::new();
    let mut input_tokens = 0u32;
    let mut output_tokens = 0u32;
    let mut finish_reason = "stop".to_string();

    // 累积 tool call 参数
    let mut pending_calls: Vec<(String, String, String)> = Vec::new(); // (index, id, name, arguments_json)
    let mut pending_index = String::new();

    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| format!("SSE 读取错误: {e}"))?;
        buffer.push_str(&String::from_utf8_lossy(&chunk));

        while let Some(pos) = buffer.find("\n\n") {
            let event_str = buffer[..pos].to_string();
            buffer = buffer[pos + 2..].to_string();

            for line in event_str.lines() {
                let line = line.trim();
                if line.is_empty() || !line.starts_with("data: ") { continue; }
                let data = &line[6..];
                if data == "[DONE]" { continue; }

                let event: Value = match serde_json::from_str(data) {
                    Ok(v) => v,
                    Err(_) => continue,
                };

                if let Some(choices) = event["choices"].as_array() {
                    for choice in choices {
                        // Reasoning delta（DeepSeek R1 思维链）
                        if let Some(rc) = choice["delta"]["reasoning_content"].as_str() {
                            if !rc.is_empty() {
                                let _ = tx.send(StreamEvent::ReasoningDelta { text: rc.to_string() }).await;
                            }
                        }
                        // Text delta
                        if let Some(text) = choice["delta"]["content"].as_str() {
                            if !text.is_empty() {
                                let _ = tx.send(StreamEvent::TextDelta { text: text.to_string() }).await;
                            }
                        }

                        // Tool calls delta
                        if let Some(tc_list) = choice["delta"]["tool_calls"].as_array() {
                            for tc in tc_list {
                                let idx = tc["index"].as_u64().map(|i| i.to_string()).unwrap_or_default();
                                let id = tc["id"].as_str().unwrap_or("").to_string();
                                let func_name = tc["function"]["name"].as_str().unwrap_or("").to_string();
                                let args = tc["function"]["arguments"].as_str().unwrap_or("").to_string();

                                if !id.is_empty() || !idx.is_empty() {
                                    // 新 tool call 或已有 call 的参数
                                    let key = if !id.is_empty() { id.clone() } else { idx.clone() };
                                    let existing = pending_calls.iter_mut().find(|(k, _, _)| *k == key);
                                    if let Some((_, _, ref mut accumulated)) = existing {
                                        accumulated.push_str(&args);
                                    } else if !func_name.is_empty() || !args.is_empty() {
                                        pending_calls.push((key, func_name, args));
                                    }
                                }
                            }
                        }

                        // Finish reason
                        if let Some(fr) = choice["finish_reason"].as_str() {
                            if !fr.is_empty() && fr != "null" {
                                finish_reason = fr.to_string();

                                // 发送累积的 tool calls
                                for (cid, cname, c_args) in pending_calls.drain(..) {
                                    let input: Value = if c_args.is_empty() {
                                        serde_json::json!({})
                                    } else {
                                        serde_json::from_str(&c_args).unwrap_or(serde_json::json!({"raw": c_args}))
                                    };
                                    let _ = tx.send(StreamEvent::ToolCallStart {
                                        id: cid,
                                        name: cname,
                                        input,
                                    }).await;
                                }
                            }
                        }
                    }
                }

                if let Some(usage) = event.get("usage") {
                    input_tokens = usage["prompt_tokens"].as_u64().unwrap_or(0) as u32;
                    output_tokens = usage["completion_tokens"].as_u64().unwrap_or(0) as u32;
                }

                // x_groq usage
                if let Some(usage) = event.get("x_groq").and_then(|g| g.get("usage")) {
                    input_tokens = usage["prompt_tokens"].as_u64().unwrap_or(0) as u32;
                    output_tokens = usage["completion_tokens"].as_u64().unwrap_or(0) as u32;
                }
            }
        }
    }

    let _ = tx.send(StreamEvent::Done {
        finish_reason: finish_reason.clone(),
        input_tokens,
        output_tokens,
    }).await;

    Ok(Usage { input_tokens, output_tokens })
}

File: src/lib/api/openai-compat.ts
/**
 * OpenAI 兼容 API provider — 原生 fetch() + SSE 流解析
 * 对齐 CC proxy: anthropicToOpenaiChat + openaiChatStreamToAnthropic
 * 支持: DeepSeek, OpenAI, Qwen, OpenRouter, Groq, Together, Mistral, Ollama 等
 */
import type { ToolDef, StreamEvent, ChatMessage, ProviderConfig, Usage } from './types'

// ====== 消息转换：stardust → OpenAI Chat Completions ======

interface OpenAIMessage {
  role: string
  content: string | null
  tool_calls?: Array<{
    id: string
    type: 'function'
    function: { name: string; arguments: string }
  }>
  tool_call_id?: string
  reasoning_content?: string
}

function convertMessages(messages: ChatMessage[], system?: string): OpenAIMessage[] {
  const result: OpenAIMessage[] = []

  // System prompt
  if (system) {
    result.push({ role: 'system', content: system })
  }

  for (const m of messages) {
    if (m.role === 'system') {
      result.push({ role: 'system', content: m.content })
    } else if (m.role === 'user') {
      result.push({ role: 'user', content: m.content })
    } else if (m.role === 'assistant') {
      const msg: OpenAIMessage = { role: 'assistant', content: m.content || '' }
      if (m.toolCalls && m.toolCalls.length > 0) {
        msg.tool_calls = m.toolCalls.map(tc => ({
          id: tc.id,
          type: 'function' as const,
          function: { name: tc.name, arguments: JSON.stringify(tc.input) },
        }))
      }
      if (m.reasoning_content) msg.reasoning_content = m.reasoning_content
      result.push(msg)
    } else if (m.role === 'tool') {
      result.push({
        role: 'tool',
        content: m.content,
        tool_call_id: m.toolCallId || '',
      })
    }
  }

  return result
}

// ====== 工具转换：stardust → OpenAI function ======

interface OpenAITool {
  type: 'function'
  function: {
    name: string
    description: string
    parameters: Record<string, unknown>
  }
}

function convertTools(tools: Record<string, ToolDef>): OpenAITool[] {
  return Object.entries(tools).map(([name, tool]) => ({
    type: 'function' as const,
    function: {
      name,
      description: tool.description,
      parameters: (tool.inputSchema || { type: 'object', properties: {} }),
    },
  }))
}

// ====== SSE 流解析 ======

interface SSEChunk {
  choices?: Array<{
    delta?: {
      content?: string | null
      reasoning_content?: string | null
      tool_calls?: Array<{
        index: number
        id?: string
        function?: { name?: string; arguments?: string }
      }>
    }
    finish_reason?: string | null
  }>
  usage?: {
    prompt_tokens?: number
    completion_tokens?: number
    prompt_cache_hit_tokens?: number
  }
}

class SSELineDecoder {
  private buffer = ''

  push(chunk: string): string[] {
    this.buffer += chunk
    const lines = this.buffer.split('\n')
    this.buffer = lines.pop() || ''
    return lines.filter(l => l.trim()).map(l => l.trim())
  }

  flush(): string[] {
    const lines = this.buffer.split('\n').filter(l => l.trim())
    this.buffer = ''
    return lines
  }
}

function parseSSEChunk(line: string): SSEChunk | null {
  if (!line.startsWith('data: ')) return null
  const json = line.slice(6)
  if (json === '[DONE]') return null
  try {
    return JSON.parse(json)
  } catch {
    return null
  }
}

// ====== finish_reason 映射 (CC 对齐) ======

function mapFinishReason(reason: string | null | undefined): string {
  switch (reason) {
    case 'stop': return 'stop'
    case 'tool_calls': return 'tool_calls'
    case 'length': return 'length'
    case 'content_filter': return 'stop'
    default: return 'stop'
  }
}

// ====== 流式主函数 ======

export async function* streamOpenAICompat(
  config: ProviderConfig,
  messages: ChatMessage[],
  tools: Record<string, ToolDef>,
  systemPrompt?: string,
  signal?: AbortSignal,
  maxOutputTokens?: number,
  thinkingBudgetTokens?: number,
): AsyncGenerator<StreamEvent> {
  const baseUrl = (config.baseUrl || 'https://api.deepseek.com').replace(/\/+$/, '')
  const url = `${baseUrl}/v1/chat/completions`

  const openaiMessages = convertMessages(messages, systemPrompt)
  const openaiTools = convertTools(tools)
  const hasTools = openaiTools.length > 0

  const body: Record<string, unknown> = {
    model: config.modelId,
    messages: openaiMessages,
    stream: true,
    stream_options: { include_usage: true },
    temperature: 0.3,
  }

  // 对齐 CC: max_tokens 剥离策略 — 不传时让模型用自身默认值
  if (maxOutputTokens != null) {
    body.max_tokens = maxOutputTokens
  }
  if (hasTools) {
    body.tools = openaiTools
  }

  // OpenAI o 系列模型支持 reasoning_effort
  if (thinkingBudgetTokens && thinkingBudgetTokens > 0) {
    const isOpenAI = config.provider === 'OpenAI' || config.baseUrl?.includes('openai.com')
    if (isOpenAI) {
      const budget = thinkingBudgetTokens
      body.reasoning_effort = budget <= 1024 ? 'low' : budget <= 4096 ? 'medium' : 'high'
    }
  }

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${config.apiKey}`,
    },
    body: JSON.stringify(body),
    signal,
  })

  if (!response.ok) {
    const text = await response.text().catch(() => '')
    console.error('[openai-compat] DeepSeek error body:', text.slice(0, 500))
    console.error('[openai-compat] Request body:', JSON.stringify(body).slice(0, 1000))
    throw new Error(`API error ${response.status}: ${text.slice(0, 200)}`)
  }

  if (!response.body) {
    throw new Error('No response body')
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  const lineDecoder = new SSELineDecoder()

  // 工具调用状态机（对齐 CC openaiChatStreamToAnthropic）
  const toolCalls: Map<number, {
    id: string
    name: string
    argsBuffer: string
  }> = new Map()

  let inputTokens = 0
  let outputTokens = 0
  let finishReason = 'stop'
  let textStarted = false
  let aborted = false
  let reasoningAccum = ''   // 累积 reasoning，用于去重
  let echoStripped = false
  let contentAccum = ''    // 累积 content（在剥离回显前不单独 yield）
  let absorbedLen = 0      // contentAccum 中已被判定为 echo 前缀的字符数
  let contentYielded = false // 是否已有 content 通过 text-delta 正常 yield
  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      if (signal?.aborted) { aborted = true; break }

      const text = decoder.decode(value, { stream: true })
      const lines = lineDecoder.push(text)

      for (const line of lines) {
        const chunk = parseSSEChunk(line)
        if (!chunk) continue

        // Usage
        if (chunk.usage) {
          inputTokens = chunk.usage.prompt_tokens ?? inputTokens
          outputTokens = chunk.usage.completion_tokens ?? outputTokens
        }

        const choice = chunk.choices?.[0]
        if (!choice) continue

        // Finish reason
        if (choice.finish_reason) {
          finishReason = mapFinishReason(choice.finish_reason)
        }

        const delta = choice.delta
        if (!delta) continue

        // Reasoning：累积并透传
        if (delta.reasoning_content) {
          reasoningAccum += delta.reasoning_content
          yield { type: 'reasoning-delta', text: delta.reasoning_content, reasoningId: 'reasoning-0' }
        }

        // Text：先累积，再与 reasoning 比对剥离回显
        // 两种 echo 模式都要处理：
        //   A) content 是 reasoning 的前缀（content 比 reasoning 短，还在收 reasoning）
        //   B) reasoning 是 content 的前缀（content 已超出 reasoning，只取超出部分）
        //   C) 部分重叠：之前有 absorbedLen，只取超出部分
        if (delta.content) {
          if (!textStarted) textStarted = true
          contentAccum += delta.content

          if (!echoStripped && reasoningAccum) {
            // 模式 A：content 短于 reasoning，是 reasoning 的前缀 → 吸收
            if (reasoningAccum.startsWith(contentAccum)) {
              absorbedLen = contentAccum.length
              console.log('[echo] A absorb:', absorbedLen, 'rLen:', reasoningAccum.length, 'cLen:', contentAccum.length)
              continue
            }
            // 模式 B：reasoning 短于 content，reasoning 是 content 的前缀 → 只取超出部分
            if (contentAccum.startsWith(reasoningAccum)) {
              echoStripped = true
              const newContent = contentAccum.slice(reasoningAccum.length)
              console.log('[echo] B content-outran rLen:', reasoningAccum.length, 'cLen:', contentAccum.length, 'yield:', newContent.length)
              if (newContent) {
                contentYielded = true; yield { type: 'text-delta', text: newContent }
              }
              continue
            }
            // 模式 C：部分重叠 → 取超出吸收长度的部分
            if (absorbedLen > 0) {
              echoStripped = true
              const newContent = contentAccum.slice(absorbedLen)
              console.log('[echo] C partial absorbed:', absorbedLen, 'cLen:', contentAccum.length, 'yield:', newContent.length)
              if (newContent) {
                contentYielded = true; yield { type: 'text-delta', text: newContent }
              }
              continue
            }
            // 模式 D：前缀不匹配，但可能同语言 echo——检查词级重叠
            // content 累积到一定长度后再判断，避免短片段误判
            if (contentAccum.length >= 20) {
              const words = (s: string) => s.toLowerCase()
                .split(/[\p{P}\p{Z}\n]+/u)
                .filter((w: string) => w.length > 1)
              const cWords = words(contentAccum)
              const rWords = new Set(words(reasoningAccum))
              if (cWords.length >= 3) {
                const overlap = cWords.filter((w: string) => rWords.has(w)).length
                const ratio = overlap / cWords.length
                if (ratio > 0.6) {
                  // 60%+ 词在 reasoning 中出现 → echo，吸收
                  echoStripped = true
                  console.log('[echo] D word-overlap rLen:', reasoningAccum.length,
                    'cLen:', contentAccum.length, 'overlapRatio:', ratio.toFixed(2), '→ absorbed')
                  continue
                }
              }
            }
            // 词级也不匹配 → 真正不是 echo
            if (contentAccum.length < 20) continue  // 内容还不够长，再等等
            echoStripped = true
            console.log('[echo] E yield-all rLen:', reasoningAccum.length, 'cLen:', contentAccum.length,
              'r[:30]:', JSON.stringify(reasoningAccum.slice(0, 30)),
              'c[:30]:', JSON.stringify(contentAccum.slice(0, 30)))
            contentYielded = true; yield { type: 'text-delta', text: contentAccum }
          } else {
            // 已剥离或无 reasoning：正常 yield delta
            contentYielded = true; yield { type: 'text-delta', text: delta.content }
          }
        }

        // Tool calls
        if (delta.tool_calls) {
          for (const tc of delta.tool_calls) {
            const idx = tc.index

            if (!toolCalls.has(idx)) {
              toolCalls.set(idx, { id: '', name: '', argsBuffer: '' })
            }

            const block = toolCalls.get(idx)!
            if (tc.id) block.id = tc.id
            if (tc.function?.name) block.name += tc.function.name
            if (tc.function?.arguments) block.argsBuffer += tc.function.arguments
          }
        }
      }
    }
  } catch (e: any) {
    if (e.name === 'AbortError' || signal?.aborted) {
      aborted = true
    } else {
      throw e
    }
  } finally {
    reader.releaseLock()
  }

  if (aborted) {
    yield { type: 'done', finishReason: 'abort', usage: { inputTokens, outputTokens } }
    return
  }

  // 流结束但还有未 yield 的 content（短回复/ModeD1误判）
  // 用 contentYielded 而非 echoStripped 判断——ModeD1 会设 echoStripped 但不 yield
  if (!contentYielded && contentAccum.length > absorbedLen) {
    const remaining = contentAccum.slice(absorbedLen)
    if (remaining) {
      yield { type: 'text-delta', text: remaining }
    }
  }

  // Flush any pending tool calls
  for (const [, block] of toolCalls) {
    if (block.id && block.name) {
      yield {
        type: 'tool-call-start',
        id: block.id,
        name: block.name,
        input: block.argsBuffer,
      }
    }
  }

  const usage: Usage = {
    inputTokens,
    outputTokens,
  }
  yield { type: 'done', finishReason, usage }
}
